controller/controller-http/src/main/java/com/pinterest/rocksplicator/controller/resource/Clusters.java

```diff
@@ -77,18 +79,27 @@ public Clusters(String zkPath,
   @GET
   @Path("/{clusterName : [a-zA-Z0-9\\-_]+}")
   @Produces(MediaType.APPLICATION_JSON)
-  public ClusterBean get(@PathParam("clusterName") String clusterName) {
+  public Response get(@PathParam("clusterName") String clusterName) {
     final ClusterBean clusterBean;
+    Result<ClusterBean> result = new Result<ClusterBean>(null);
     try {
       clusterBean = checkExistenceAndGetClusterBean(clusterName);
     } catch (Exception e) {
-      LOG.error("Failed to read from zookeeper.", e);
-      throw new WebApplicationException(e);
+      String message = String.format("Failed to read from zookeeper: %s", e);
+      LOG.error(message);
+      return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500)
+                     .entity(result.setMessage(message))
+                     .build();
     }
     if (clusterBean == null) {
-      throw new WebApplicationException(HttpStatus.NOT_FOUND_404);
+      String message = String.format("Znode %s does not exist or failed to parse config", clusterName);
+      return Response.status(HttpStatus.BAD_REQUEST_400)
```

Comment:
`HttpStatus.NOT_FOUND_404` means a server resource does not exist. In this case the server resource exists, just the element we need to find does not exist, therefore `400`.

you can check here: https://stackoverflow.com/questions/25378624/should-a-restful-api-return-400-or-404-when-passed-an-invalid-id